File: hash_set.py

""" 
Time Complexity : add/remove/contains - O(1)
Space Complexity : add/remove/contains - O(1) - Amortized
Did this code successfully run on Leetcode: Yes
Any problem you faced while coding this: No
"""

class MyHashSet:
    # limit = 1,000,0000
    def __init__(self):
        self.totalBuckets = 1000
        self.totalBucketItems = 1000
        self.storage = [0] * self.totalBucketItems

    # hash
    def findBucketNumber(self, key:int) -> int:
        return key % self.totalBuckets 

    def findBucketItemIdx(self, key:int) -> int:
        return key // self.totalBucketItems

    def add(self, key:int) -> None: 
        bucketNumber = self.findBucketNumber(key)
        if self.storage[bucketNumber] == 0:
            if bucketNumber == 0:
                # 1 additional spot
                self.storage[bucketNumber] = [0] * (self.totalBucketItems + 1)
            else:
                self.storage[bucketNumber] = [0] * self.totalBucketItems
        bucketItemIdx = self.findBucketItemIdx(key)
        # print('Key: ', key, 'Add: ', bucketNumber, bucketItemIdx, self.storage[bucketNumber][bucketItemIdx])
        self.storage[bucketNumber][bucketItemIdx] = True

    def remove(self, key:int) -> None: 
        bucketNumber = self.findBucketNumber(key)
        if self.storage[bucketNumber] == 0: 
            return None
        bucketItemIdx = self.findBucketItemIdx(key)
        # print('Key: ', key, 'Remove: ', bucketNumber, bucketItemIdx, self.storage[bucketNumber][bucketItemIdx])
        self.storage[bucketNumber][bucketItemIdx] = False
            

    def contains(self, key:int) -> bool:
        bucketNumber = self.findBucketNumber(key)
        if self.storage[bucketNumber] == 0:
            return False
        bucketItemIdx = self.findBucketItemIdx(key)
        # print('Key: ', key, 'Contains: ', bucketNumber, bucketItemIdx, self.storage[bucketNumber][bucketItemIdx])
        return self.storage[bucketNumber][bucketItemIdx]

File: hash_set.js
// Time Complexity : O(1)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Yes, while creating an initial storage bucket array

// Your code here along with comments explaining your approach

var MyHashSet = function () {
  this.primaryBucket = 1000;
  this.secondaryBucket = 1000;
  this.storage = this.createStorage(this.primaryBucket, this.secondaryBucket);
};

MyHashSet.prototype.createStorage = function (rows, columns) {
  let array = new Array(rows);

  for (let i = 0; i < rows; i++) {
    array[i] = new Array(columns).fill(false); // Fill each inner array with 'false'
  }

  return array;
};

MyHashSet.prototype.getPrimaryIndex = function (key) {
  return key % this.storage.length;
};

MyHashSet.prototype.getSecondaryIndex = function (key) {
  return key / this.storage.length;
};

/**
 * @param {number} key
 * @return {void}
 */
MyHashSet.prototype.add = function (key) {
  let primaryIndex = this.getPrimaryIndex(key);
  if (this.storage[primaryIndex] == null) {
    this.storage[primaryIndex] = [];
  }
  let secondaryIndex = this.getSecondaryIndex(key);
  this.storage[primaryIndex][secondaryIndex] = true;
};

/**
 * @param {number} key
 * @return {void}
 */
MyHashSet.prototype.remove = function (key) {
  let primaryIndex = this.getPrimaryIndex(key);
  if (this.storage[primaryIndex] == null) {
    return;
  }
  let secondaryIndex = this.getSecondaryIndex(key);
  this.storage[primaryIndex][secondaryIndex] = false;
};

/**
 * @param {number} key
 * @return {boolean}
 */
MyHashSet.prototype.contains = function (key) {
  let primaryIndex = this.getPrimaryIndex(key);
  if (this.storage[primaryIndex] == null) {
    return false;
  }
  let secondaryIndex = this.getSecondaryIndex(key);
  return this.storage[primaryIndex][secondaryIndex] || false;
};

test("Scenario #1:", () => {
  var obj = new MyHashSet();
  obj.add(1001);
  obj.add(1006);
  expect(obj.contains(1001)).toStrictEqual(true);
  expect(obj.contains(1006)).toStrictEqual(true);
});
